File: scripts/youngcha/distribution/cruciblepump.js
const rotL = require("libraries/rotpowerlib");
const heatlib = require("libraries/heatlib");
const graphLib = require("libraries/graphlib");
const crucibleLib = require("libraries/cruciblelib");

const fillAm = [1.0,0.5,0.25];

let cpblankobj = graphLib.init();
graphLib.addGraph(cpblankobj, crucibleLib.baseTypes.crucibleMultiConnector);
graphLib.addGraph(cpblankobj, heatlib.baseTypesHeat.heatConnector);
const cruciblePump = graphLib.finaliseExtend(Block, Building,"crucible-pump",cpblankobj,{
	load(){
		this.super$load();
		this.top = [Core.atlas.find(this.name+"-top1"),Core.atlas.find(this.name+"-top2"),Core.atlas.find(this.name+"-top3"),Core.atlas.find(this.name+"-top4")];
		this.liquidSprite = Core.atlas.find(this.name+"-liquid");
		this.heatSprite = Core.atlas.find(this.name+"-heat");
		this.bottom = Core.atlas.find(this.name+"-bottom");
	},
},{
	filterItem:null,
	pumpMode:0,
	flowRate:0,
	flowAnimation:0,
	
	getFilterItemConfig(){return this.filterItem;},
	getPumpModeConfig(){return this.pumpMode;},
	
	buildConfiguration(table) {
		let that= this;
		table.labelWrap("Fill until:").growX().pad(5).center();
		table.row();
		table.table(cons((btable)=>{
			btable.button("Full", Styles.clearPartialt, run(() => {that.configure(0);})).left().size(50).disabled(boolf(b =>{ return that.pumpMode==0 ;}));
			btable.button("50%", Styles.clearPartialt, run(() => {that.configure(1);})).left().size(50).disabled(boolf(b =>{ return that.pumpMode==1 ;}));
			btable.button("25%", Styles.clearPartialt, run(() => {that.configure(2);})).left().size(50).disabled(boolf(b =>{ return that.pumpMode==2 ;}));
		}));
		
		
		table.row();
		table.labelWrap("Pump:").growX().pad(5).center();
		table.row();
		ItemSelection.buildTable(table, Vars.content.items(), () => this.filterItem, cons((x)=>{this.configure(x);}));
		table.setBackground(Styles.black5);
	},
	configured(player, value) {
		print(player);
		
		if(value instanceof Item){
			this.filterItem = value;
			print("set item: "+value);	
		}else if(!isNaN(value)){
			this.pumpMode = value;
			print("set mode: "+value);	
		}
	},
	displayExt(table){
		let that =this;
		let ps = " " + StatUnit.perSecond.localized();
		table.row();
        table.table(
            cons(sub => {
                sub.clearChildren();
                sub.left();
				if(that.filterItem){
					sub.image(that.filterItem.icon(Cicon.medium));
					sub.label(prov(() => {
						return Strings.fixed(that.flowRate *10.0, 2) + "units" + ps;
					})).color(Color.lightGray);
				}else{
					sub.labelWrap("No filter selected").color(Color.lightGray);
				}
            })
        ).left();
	},
	updatePost(){
		let rate = 0.08;
		let dex = this.getGraphConnector("crucible graph");
		this.flowRate/=2.0;
		if(this.filterItem){
			let fromNet = dex.getNetworkFromSet(1);
			let toNet = dex.getNetworkFromSet(0);
			if(fromNet&&toNet&&fromNet.contains&&toNet.contains){
				for(let i =0;i<fromNet.contains.length;i++){
					let fnc = fromNet.contains[i];
					if(fnc.item!=this.filterItem){
						continue;
					}
					let transfer = Math.min(rate*this.edelta(),fnc.volume*fnc.meltedRatio);
					transfer = Math.min(toNet.getRemainingSpace(),transfer);
					let toG = toNet.getMeltFromID(fnc.id);
					if(toG){
						transfer = Math.min(toNet.getLiquidCapacity()*fillAm[this.pumpMode]-toG.volume,transfer);
					}
					if(transfer<0){
						break;
					}
					if(transfer){
						fromNet.addLiquidToSlot(fnc,-transfer);
						toNet.addMeltItem(crucibleLib.meltTypes[fnc.id],transfer,true);
						this.flowRate = transfer;
					}
					break;
				}
				
			}
		}
		this.flowAnimation+=this.flowRate*0.4;
	},
	draw() {
		let temp = this.getGraphConnector("heat graph").getTemp();
		Draw.rect(cruciblePump.bottom, this.x, this.y, 0);
		if(this.filterItem){
			Draw.color(this.filterItem.color, Mathf.clamp(this.flowRate*60.0));	
			rotL.drawSlideRect(cruciblePump.liquidSprite,this.x, this.y,16,16, 32,16, this.rotdeg()+180,16,this.flowAnimation);
			Draw.color();
		}
		heatlib.drawHeat(cruciblePump.heatSprite, this.x, this.y, this.rotdeg(), temp);
		Draw.rect(cruciblePump.top[this.rotation], this.x, this.y, 0);
        this.drawTeamTop();
	},
	writeExt(stream) {
		stream.s(this.filterItem == null ? -1 : this.filterItem.id);
		stream.i(this.pumpMode);
	},
	readExt(stream, revision) {
		this.filterItem = Vars.content.item(stream.s());
		this.pumpMode = stream.i();
	}
});
cruciblePump.rotate = true;
cruciblePump.update = true;
cruciblePump.solid = true;
cruciblePump.configurable = true;
cruciblePump.consumes.power(1.0);
cruciblePump.getGraphConnectorBlock("crucible graph").setAccept([1,1,0,0,2,2,0,0]);
cruciblePump.getGraphConnectorBlock("crucible graph").setDoesCrafting(false);
cruciblePump.getGraphConnectorBlock("crucible graph").setBaseLiquidCapacity(10);

cruciblePump.getGraphConnectorBlock("heat graph").setAccept( [1,1,1,1,1,1,1,1]);
cruciblePump.getGraphConnectorBlock("heat graph").setBaseHeatConductivity(0.1);
cruciblePump.getGraphConnectorBlock("heat graph").setBaseHeatCapacity(50);
cruciblePump.getGraphConnectorBlock("heat graph").setBaseHeatRadiativity(0.003);

File: main/src/unity/world/blocks/distribution/CruciblePump.java
package unity.world.blocks.distribution;

import arc.*;
import arc.graphics.*;
import arc.graphics.g2d.*;
import arc.math.*;
import arc.scene.ui.layout.*;
import arc.util.*;
import arc.util.io.*;
import mindustry.*;
import mindustry.type.*;
import mindustry.ui.*;
import mindustry.world.blocks.*;
import mindustry.world.meta.*;
import unity.graphics.*;
import unity.world.blocks.*;
import unity.world.graph.*;
import unity.world.meta.*;
import unity.world.modules.*;

import static mindustry.Vars.*;

public class CruciblePump extends GraphBlock{
    public static final float[] fillAm = new float[]{1f, 0.5f, 0.25f};
    public final TextureRegion[] topRegions = new TextureRegion[4];
    public TextureRegion bottomRegion;

    public CruciblePump(String name){
        super(name);

        rotate = solid = configurable = true;
        config(Item.class, (CruciblePumpBuild build, Item item) -> build.filterItem = item);
        config(Integer.class, (CruciblePumpBuild build, Integer value) -> {
            unity.Unity.print(value, value & 3, value >>> 2);
            build.pumpMode = value & 3;
            if(value > 2) build.filterItem = Vars.content.item(value >>> 2);
        });
        configClear((CruciblePumpBuild build) -> build.filterItem = null);
    }

    @Override
    public void load(){
        super.load();

        for(int i = 0; i < 4; i++) topRegions[i] = Core.atlas.find(name + "-top" + (i + 1));
        bottomRegion = Core.atlas.find(name + "-bottom");
    }

    public class CruciblePumpBuild extends GraphBuild{
        Item filterItem;
        float flowRate, flowAnimation;
        int pumpMode;

        @Override
        public void buildConfiguration(Table table){
            table.labelWrap("Fill until:").growX().pad(5f).center().row();
            table.table(bt -> {
                bt.button("Full", Styles.clearPartialt, () -> configure(0)).left().size(50f).disabled(b -> pumpMode == 0);
                bt.button("50%", Styles.clearPartialt, () -> configure(1)).left().size(50f).disabled(b -> pumpMode == 1);
                bt.button("25%", Styles.clearPartialt, () -> configure(2)).left().size(50f).disabled(b -> pumpMode == 2);
            }).row();

            table.labelWrap("Pump:").growX().pad(5f).center().row();
            ItemSelection.buildTable(table, Vars.content.items(), () -> filterItem, this::configure);
            table.setBackground(Styles.black5);
        }

        @Override
        public void displayExt(Table table){
            String ps = " " + StatUnit.perSecond.localized();
            table.row();
            table.table(sub -> {
                sub.clearChildren();
                sub.left();
                if(filterItem != null){
                    sub.image(filterItem.uiIcon).size(iconMed);
                    sub.label(() -> Strings.fixed(flowRate * 10f, 2) + "units" + ps).color(Color.lightGray);
                }else{
                    sub.labelWrap("No filter selected").color(Color.lightGray);
                }
            }).left();
        }

        @Override
        public void updatePost(){
            float rate = 0.08f;
            GraphCrucibleModule dex = crucible();
            flowRate /= 2f;

            if(filterItem != null){
                CrucibleGraph fromNet = dex.getNetworkFromSet(1);
                CrucibleGraph toNet = dex.getNetworkFromSet(0);

                if(fromNet != null && toNet != null){
                    for(var fnc : fromNet.contains()){
                        if(fnc.item != filterItem) continue;

                        float transfer = Math.min(toNet.getRemainingSpace(), Math.min(rate * edelta(), fnc.volume * fnc.meltedRatio));
                        CrucibleData toG = toNet.getMeltFromID(fnc.id);

                        if(toG != null) transfer = Math.min(toNet.totalCapacity() * fillAm[pumpMode] - toG.volume, transfer);
                        if(transfer <= 0f) break;

                        fromNet.addLiquidToSlot(fnc, -transfer);
                        toNet.addMeltItem(MeltInfo.all[fnc.id], transfer, true);
                        flowRate = transfer;

                        break;
                    }
                }
            }
            flowAnimation += flowRate * 0.4f;
        }

        @Override
        public void draw(){
            Draw.rect(bottomRegion, x, y);
            if(filterItem != null){
                Draw.color(filterItem.color, Mathf.clamp(flowRate * 60f));
                UnityDrawf.drawSlideRect(liquidRegion, x, y, 16f, 16f, 32f, 16f, rotdeg() + 180f, 16, flowAnimation);

                Draw.color();
            }

            UnityDrawf.drawHeat(heatRegion, x, y, rotdeg(), heat().getTemp());
            Draw.rect(topRegions[rotation], x, y);

            drawTeamTop();
        }

        @Override
        public void writeExt(Writes write){
            write.s(filterItem == null ? -1 : filterItem.id);
            write.b(pumpMode);
        }

        @Override
        public void readExt(Reads read, byte revision){
            filterItem = Vars.content.item(read.s());
            pumpMode = read.b();
        }

        @Override
        public Integer config(){
            return pumpMode + (filterItem != null ? (filterItem.id + 1) << 2 : 0);
        }
    }
}
